core/src/main/java/org/apache/calcite/sql/fun/BodoSqlTimestampAddFunction.java

```diff
@@ -0,0 +1,372 @@
+
+import org.apache.calcite.avatica.util.TimeUnit;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.sql.SqlCallBinding;
+import org.apache.calcite.sql.SqlFunction;
+import org.apache.calcite.sql.SqlFunctionCategory;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.type.OperandTypes;
+import org.apache.calcite.sql.type.SqlReturnTypeInference;
+import org.apache.calcite.sql.type.SqlTypeFamily;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.tools.ValidationException;
+
+import java.util.Locale;
+
+import static org.apache.calcite.sql.validate.SqlNonNullableAccessors.getOperandLiteralValueOrThrow;
+import static org.apache.calcite.util.Static.RESOURCE;
+
+/**
+ * The <code>TIMESTAMPADD</code> function, which adds an interval to a
+ * datetime (TIMESTAMP, TIME or DATE).
+ *
+ * <p>The SQL syntax is
+ *
+ * <blockquote>
+ * <code>TIMESTAMPADD(<i>timestamp interval</i>, <i>quantity</i>,
+ * <i>datetime</i>)</code>
+ * </blockquote>
+ * Accepts both quoted and unquoted time units now.
+ *
+ * <p>Returns modified datetime.
+ */
+public class BodoSqlTimestampAddFunction extends SqlFunction {
+
+  private static final int MILLISECOND_PRECISION = 3;
+  private static final int MICROSECOND_PRECISION = 6;
+
+  private static final SqlReturnTypeInference RETURN_TYPE_INFERENCE =
+      opBinding -> {
+        final RelDataTypeFactory typeFactory = opBinding.getTypeFactory();
+        assert opBinding instanceof SqlCallBinding;
+        SqlCallBinding opBindingWithCast = (SqlCallBinding) opBinding;
+        RelDataType arg0Type = opBindingWithCast.getOperandType(0);
+        TimeUnit arg0timeUnit;
+        switch (arg0Type.getSqlTypeName()) {
+          // This must be a constant string or time unit input,
+          // due to the way that we handle the parsing
+        case CHAR:
+        case VARCHAR:
+          //This will fail if the value is a non-literal
+          try {
+            arg0timeUnit = standardizeTimeUnit("TIMESTAMPADD",
+              opBindingWithCast.getOperandLiteralValue(0, String.class),
+       opBindingWithCast.getOperandType(2).getSqlTypeName() == SqlTypeName.TIME);
+          } catch (Throwable e) {
+            throw opBindingWithCast.getValidator().newValidationError(opBindingWithCast.getCall(),
+                RESOURCE.wrongTimeUnit("TIMESTAMPADD", e.toString()));
+          }
+          break;
+
+        default:
+          arg0timeUnit = getOperandLiteralValueOrThrow(opBinding, 0, TimeUnit.class);
+        }
+
+        RelDataType ret;
+        try {
+          ret = deduceType(typeFactory, arg0timeUnit,
+              opBinding.getOperandType(1), opBinding.getOperandType(2));
+        } catch (Throwable e) {
+          throw opBindingWithCast.getValidator().newValidationError(opBindingWithCast.getCall(),
+              RESOURCE.wrongTimeUnit("TIMESTAMPADD", e.toString()));
```

Comment:
If we throw a runtime error due to the case that the input types are wrong, than this error message will be wrong. We likely need to case on the thrown error.

When this is done, can you add an error message test for incorrect input types, similar to the others?